source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,512 @@
+  // The phi should only have two incoming value pairs.
+  if (phi->NumInOperands() != 4) {
+    return CreateCantComputeNode();
+  }
+
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+
+  // Get the basic block this instruction belongs to.
+  ir::BasicBlock* basic_block =
+      context_->get_instr_block(const_cast<ir::Instruction*>(phi));
+
+  // And then the function that the basic blocks belongs to.
+  ir::Function* function = basic_block->GetParent();
+
+  // Use the function to get the loop descriptor.
+  ir::LoopDescriptor* loop_descriptor = context_->GetLoopDescriptor(function);
+
+  // We only handle phis in loops at the moment.
+  if (!loop_descriptor) return CreateCantComputeNode();
+
+  // Get the innermost loop which this block belongs to.
+  ir::Loop* loop = (*loop_descriptor)[basic_block->id()];
+
+  // If the loop doesn't exist or doesn't have a preheader or latch block, exit
+  // out.
+  if (!loop || !loop->GetLatchBlock() || !loop->GetPreHeaderBlock() ||
+      loop->GetHeaderBlock() != basic_block)
+    return CreateCantComputeNode();
+
+  std::unique_ptr<SERecurrentNode> phi_node{new SERecurrentNode(this, loop)};
+  recurrent_node_map_[phi] = phi_node.get();
+
+  for (uint32_t i = 0; i < phi->NumInOperands(); i += 2) {
+    uint32_t value_id = phi->GetSingleWordInOperand(i);
+    uint32_t incoming_label_id = phi->GetSingleWordInOperand(i + 1);
+
+    ir::Instruction* value_inst = def_use->GetDef(value_id);
+    SENode* value_node = AnalyzeInstruction(value_inst);
+
+    if (incoming_label_id == loop->GetPreHeaderBlock()->id()) {
+      phi_node->AddOffset(value_node);
+    } else if (incoming_label_id == loop->GetLatchBlock()->id()) {
+      // Assumed to be in the form of step + phi.
+      if (value_node->GetType() != SENode::Add) return CreateCantComputeNode();
+
+      SENode* step_node = nullptr;
+      SENode* phi_operand = nullptr;
+      SENode* operand_1 = value_node->GetChild(0);
+      SENode* operand_2 = value_node->GetChild(1);
+
+      // Find which node is the step term.
+      if (!operand_1->AsSERecurrentNode())
+        step_node = operand_1;
+      else if (!operand_2->AsSERecurrentNode())
+        step_node = operand_2;
+
+      // Find which node is the recurrent expression.
+      if (operand_1->AsSERecurrentNode())
+        phi_operand = operand_1;
+      else if (operand_2->AsSERecurrentNode())
+        phi_operand = operand_2;
+
+      // If it is not in the form step + phi exit out.
+      if (!(step_node && phi_operand)) return CreateCantComputeNode();
+
+      // If the phi operand is not the same phi node exit out.
+      if (phi_operand != phi_node.get()) return CreateCantComputeNode();
+
+      if (!IsLoopInvariant(loop, step_node)) {
+        return CreateCantComputeNode();
+      }
+
+      phi_node->AddCoefficient(step_node);
+    }
+  }
+
+  recurrent_node_map_[phi] = GetCachedOrAdd(std::move(phi_node));
+
+  return recurrent_node_map_[phi];
```

Comment:
Better to update the cache in ScalarEvolutionAnalysis::AnalyzeInstruction().  Alternately, it seems that only Phi instructions are mapped here.  Why not the others?  Perhaps the field needs a different name?

Reply:
A different name from `recurrent_node_map_` or was it still `instruction_map_` in the version reviewed? It was confusingly named before but the purpose is to break the recursion inherent to phi nodes, as analysing the step will lead back to analysing the phi again so we look up that map in the `AnalyzeInstruction` function.